[rustdoc] Add sans-serif font setting

GuillaumeGomez opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/52449.

This PR adds a new setting introducing the possibility to switch to a sans-serif font (`Fira Sans`) for the text.

Can be tested [here](https://rustdoc.crud.net/imperio/sans-serif/std/index.html).

cc @rust-lang/rustdoc-frontend 
r? @notriddle

notriddle commented:
If you want Fira Sans for the main body type, then [Fira Mono](https://github.com/mozilla/Fira/releases/tag/4.106) should be used for code. That way they have consistent point sizes, x-height, and other features.

GuillaumeGomez commented:
Sounds good to me!

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12087086300/job/33707800791) [(plain)](https://github.com/rust-lang/rust/commit/43681bf51d240b8097bbb2b09e9b54206e5d684a/checks/33707800791/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.0s done
#16 DONE 32.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.96s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
---- [run-make] tests/run-make/emit-shared-files stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-18/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-18/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TARGET_RPATH_ENV="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/rmake_out:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/rmake"
--- stderr -------------------------------
thread 'main' panicked at /checkout/tests/run-make/emit-shared-files/rmake.rs:48:5:
assertion `left == right` failed
  left: 0
```

</details>

GuillaumeGomez commented:
Checking what studies say about it.

 * https://dergipark.org.tr/en/download/article-file/252226 : this one seems to say that sans serif fonts are more adapted for computers and improve readability (although they only seemed to compare `Verdana` and `Times New Roman`)
 * https://eric.ed.gov/?id=EJ1067757 : this one says that there is no significant difference between both
 * https://www.researchgate.net/profile/Ricardo-Baeza-Yates/publication/262320823_Good_fonts_for_dyslexia/links/628133b3107cae2919a8b26d/Good-fonts-for-dyslexia.pdf: this one seems to say that sans serif are better
 * https://library.oapen.org/bitstream/handle/20.500.12657/53344/978-3-030-90984-0.pdf : no difference
 * https://peerj.com/articles/cs-1139.pdf : no difference
 * https://www.researchgate.net/profile/Filiz-Cicek/publication/304358587_How_Serif_and_Sans_Serif_Typefaces_Influence_Reading_on_Screen_An_Eye_Tracking_Study/links/5783769608ae37d3af6bedaf/How-Serif-and-Sans-Serif-Typefaces-Influence-Reading-on-Screen-An-Eye-Tracking-Study.pdf : sans serif is a bit better

So seems like there is no difference or it's a bit better in favour of sans serif. In any case I want to ensure the team is on board with this as it's a big UI change. Better take our time to ensure that we're not missing something before approving the change.

notriddle commented:
I realized I was kind of terse. Here's a slightly more detailed opinion, along with the reason why I hold it:

1. If you choose a body font, you need to choose a code font that goes with it, because Rustdoc mixes them freely in layouts like this one:

   | before | after |
   |--|--|
   ![image](https://github.com/user-attachments/assets/659cc208-50f4-4fa6-9780-96e79b8e656e) |![image](https://github.com/user-attachments/assets/449b7485-ae35-432d-b217-bce0672d8bb2)

   If you're not sure why I think "before" looks better, focus on the phrase "or target".

2. If we change something, we need to address the reasons why it was done that way originally. It's fine if the reason seems to be "nobody thought of it," but the choice of a serif font here was definitely deliberate.

   As a user with bad enough vision that they won't let me drive a car, Fira Sans doesn't seem significantly more readable than Source Serif. This adds up, since, even though it's a serif font, [Source Serif is supposed to be usable as a screen font](https://blog.typekit.com/2014/05/20/source-serif-pro/). On the other hand, cataracts, astigmatism, and amblyopia are all "bad eyesight," but they have totally different effects. Just because it WorksForMe(tm) doesn't mean it works for anybody else.

I think this is why even relatively un-customizable apps like Apple Books have multiple font packs built in. It's few enough that they can make sure every one of them looks good, but diverse enough to satisfy the most common problems someone might have with extended reading.

I think we should add an option to the Settings panel to change the font, just like Pocket and Firefox's Reader Mode. This is because, when you consider the "bikeshedding effect," there really aren't that many people complaining about Source Serif. 

![Screenshot_2024-11-29_10-57-04](https://github.com/user-attachments/assets/ac5b17e7-6dfc-4175-b37a-693ee177a39a)

![Screenshot_2024-11-29_11-14-05](https://github.com/user-attachments/assets/c1e55273-a2d9-42b4-833c-511bc4bc50a1)

![Screenshot_2024-11-29_11-37-49](https://github.com/user-attachments/assets/fed63bc2-7b88-4c76-a545-aaee9bd15cb6)

Now that CSS Variables are being used, implementing this shouldn't actually be that hard.

bors commented:
:umbrella: The latest upstream changes (presumably #133634) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez committed:
Add new setting allowing to switch to sans serif fonts

GuillaumeGomez committed:
Add GUI test for new "sans serif fonts" setting

GuillaumeGomez commented:
I added a new setting to use "sans serif font":

![image](https://github.com/user-attachments/assets/f2ed0d6d-18d9-46cb-a719-12a347972c4d)

I updated online demo as well.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12142260309/job/33856320772) [(plain)](https://github.com/rust-lang/rust/commit/88d45d0e4f48c2a2110a7504e3ca6c0cf1eb62f1/checks/33856320772/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.232 Building wheels for collected packages: reuse
#13 3.233   Building wheel for reuse (pyproject.toml): started
#13 3.490   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.491   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.492   Stored in directory: /tmp/pip-ephem-wheel-cache-srmtqzsp/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.495 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.895 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.895 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.452 Collecting virtualenv
#13 4.452 Collecting virtualenv
#13 4.578   Downloading virtualenv-20.28.0-py3-none-any.whl (4.3 MB)
#13 4.791      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 20.5 MB/s eta 0:00:00
#13 4.832 Collecting distlib<1,>=0.3.7
#13 4.864   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.872      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 82.4 MB/s eta 0:00:00
#13 4.908 Collecting filelock<4,>=3.12.2
#13 4.940   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.974 Collecting platformdirs<5,>=3.9.1
#13 5.006   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 5.089 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.272 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.28.0
#13 DONE 5.4s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      298944 kB
DirectMap2M:     7041024 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2024-11-27/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5699 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: binary checked into source: /checkout/src/librustdoc/html/static/fonts/FiraMono-Medium.woff2
tidy error: binary checked into source: /checkout/src/librustdoc/html/static/fonts/FiraMono-Regular.woff2
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Dec  3 15:05:22 UTC 2024
  network time: Tue, 03 Dec 2024 15:05:22 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GuillaumeGomez committed:
Fix tidy errors

GuillaumeGomez committed:
Fix tidy errors

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12142575851/job/33857463130) [(plain)](https://github.com/rust-lang/rust/commit/60e8aef7741378716bb30e777e3a6b98f57de316/checks/33857463130/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
The existing /checkout/license-metadata.json file is out of date.
Run ./x run collect-license-metadata to update it.
Error: The existing /checkout/license-metadata.json file doesn't match what REUSE reports.
loading existing license information
Command has failed. Rerun with -v to see more details.
  local time: Tue Dec  3 15:46:44 UTC 2024
  network time: Tue, 03 Dec 2024 15:46:44 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12161862096/job/33917447907) [(plain)](https://github.com/rust-lang/rust/commit/247aece73aecf557b3e9d41a07797e19cd7f9131/checks/33917447907/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
gathering license information from REUSE (this might take a minute...)
finished gathering the license information from REUSE in 96.10s
loading existing license information
The existing /checkout/license-metadata.json file is out of date.
Run ./x run collect-license-metadata to update it.
Error: The existing /checkout/license-metadata.json file doesn't match what REUSE reports.
Command has failed. Rerun with -v to see more details.
  local time: Wed Dec  4 15:01:57 UTC 2024
  network time: Wed, 04 Dec 2024 15:01:57 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12162539964/job/33919734753) [(plain)](https://github.com/rust-lang/rust/commit/b1c60a6f257f42d8c0ba6e47e5589633c8969b23/checks/33919734753/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
gathering license information from REUSE (this might take a minute...)
finished gathering the license information from REUSE in 99.23s
loading existing license information
The existing /checkout/license-metadata.json file is out of date.
Run ./x run collect-license-metadata to update it.
Error: The existing /checkout/license-metadata.json file doesn't match what REUSE reports.
Command has failed. Rerun with -v to see more details.
  local time: Wed Dec  4 15:40:13 UTC 2024
  network time: Wed, 04 Dec 2024 15:40:13 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GuillaumeGomez commented:
Yeah ok, I officially have no idea what's wrong with this script.

notriddle commented on license-metadata.json:
```diff
@@ -205,53 +161,6 @@
-              ],
-              "spdx": "Apache-2.0 OR CC-BY-SA-4.0 OR MIT"
-            },
-            "name": "src/doc/embedded-book",
```
I assume it's because you've got your config.toml set up to avoid building the embedded book and gcc, so they don't wind up included in this file even though they should be.

To fix this, would either do a truly complete rust build, or I would just hand-patch in the few lines that are added in your PR.

GuillaumeGomez replied:
Ok, that makes more sense now. Thanks a lot!

notriddle commented:
The PR description and title don't describe what it does any more. That should probably be changed, then it can be more easily FCP-ed.

GuillaumeGomez commented:
Updated description. Not on computer and forgor FCP command so I'll let you start it.

notriddle commented:
@rfcbot fcp merge

GuillaumeGomez commented:
I think I checked but I'll check another time to ensure I didn't miss it.

tspiteri commented:
I think if Fira Sans is going to be an option for the text, the italic version should be included as well, otherwise you get fake italics like for example in the f32 documentation: (from test site with sans font selected)

![image](https://github.com/user-attachments/assets/07939518-a74e-4eb6-a2fc-742556c08f53)

I would expect to see something like this instead:

![image](https://github.com/user-attachments/assets/c071fc6a-7b13-400e-960b-1f7916c9edb5)

Note that for some reason, it appears that on the test site even the code is using fake italics when sans fonts are selected, whereas proper italics are used for both text and code when using the serif fonts: (from test site with sans font not selected)

![image](https://github.com/user-attachments/assets/a4a5b721-23a7-4772-b166-f25dcb1749d7)

notriddle commented:
@rfcbot concern italics

GuillaumeGomez commented:
Just checked: there is only one license for all Fira fonts from the zip I downloaded. So I think we're good on that side. Adding the italic version as well.

GuillaumeGomez committed:
Add italic for newly added sans serif fonts

GuillaumeGomez commented:
Added the italic fonts for sans serif and updated online docs.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12847583878/job/35824043255) [(plain)](https://github.com/rust-lang/rust/commit/88329534e29936a3967111cb7b169d4e395cc8fe/checks/35824043255/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
The existing /checkout/license-metadata.json file is out of date.
Run ./x run collect-license-metadata to update it.
Error: The existing /checkout/license-metadata.json file doesn't match what REUSE reports.
loading existing license information
Command has failed. Rerun with -v to see more details.
  local time: Sat Jan 18 21:58:23 UTC 2025
  network time: Sat, 18 Jan 2025 21:58:23 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

notriddle commented:
@rfcbot resolved italics

notriddle commented:
@rfcbot resolve italics

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12852904757/job/35835456994) [(plain)](https://github.com/rust-lang/rust/commit/1b6d0cf5f838ca4b3e21b9b5864589758e21e63b/checks/35835456994/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.........F (130/136)
FFFFFF     (136/136)


/checkout/tests/rustdoc-gui/enum-variants.goml enum-variants... FAILED
[ERROR] `tests/rustdoc-gui/enum-variants.goml` line 7: The following errors happened (for selector `details.non-exhaustive > summary`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "details.non-exhaustive > summary",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},

/checkout/tests/rustdoc-gui/label-next-to-symbol.goml label-next-to-symbol... FAILED
/checkout/tests/rustdoc-gui/label-next-to-symbol.goml label-next-to-symbol... FAILED
[ERROR] `tests/rustdoc-gui/label-next-to-symbol.goml` line 31: comparison didn't fail: for command `compare-elements-position-false: (
    "dt .stab.deprecated",
    "dt .stab.portability",
    ["y"],

/checkout/tests/rustdoc-gui/module-items-font.goml module-items-font... FAILED
/checkout/tests/rustdoc-gui/module-items-font.goml module-items-font... FAILED
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 3: The following errors happened (for selector `.item-table dt > a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    ".item-table dt > a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},
)`
)`
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 15: The following errors happened (for selector `#modules + .item-table dt a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#modules + .item-table dt a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},
)`
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 24: The following errors happened (for selector `#structs + .item-table dt a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#structs + .item-table dt a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},
)`
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 33: The following errors happened (for selector `#enums + .item-table dt a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#enums + .item-table dt a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},
)`
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 42: The following errors happened (for selector `#traits + .item-table dt a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#traits + .item-table dt a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},
)`
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 51: The following errors happened (for selector `#functions + .item-table dt a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#functions + .item-table dt a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},
)`
[ERROR] `tests/rustdoc-gui/module-items-font.goml` line 60: The following errors happened (for selector `#keywords + .item-table dt a`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#keywords + .item-table dt a",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},

/checkout/tests/rustdoc-gui/notable-trait.goml notable-trait... FAILED
/checkout/tests/rustdoc-gui/notable-trait.goml notable-trait... FAILED
[ERROR] `tests/rustdoc-gui/notable-trait.goml` line 52
    from `tests/rustdoc-gui/notable-trait.goml` line 64: The following errors happened: [different X values: 474.5625 (or 475) != 468]: for command `assert-position: (
            "//*[@class='tooltip popover']",
            {"x": |popover_x|}

/checkout/tests/rustdoc-gui/scrape-examples-fonts.goml scrape-examples-fonts... FAILED
/checkout/tests/rustdoc-gui/scrape-examples-fonts.goml scrape-examples-fonts... FAILED
[ERROR] `tests/rustdoc-gui/scrape-examples-fonts.goml` line 6: Error: The following CSS properties still don't match: [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `wait-for-css: (".scraped-example-title", {"font-family": |font|})`
/checkout/tests/rustdoc-gui/search-result-display.goml search-result-display... FAILED
/checkout/tests/rustdoc-gui/search-result-display.goml search-result-display... FAILED
[WARNING] `tests/rustdoc-gui/search-result-display.goml` line 37: Delta is 0 for "x", maybe try to use `compare-elements-position` instead?

[ERROR] `tests/rustdoc-gui/search-result-display.goml` line 53: The following errors happened (for selector `#crate-search`): [expected `159px` for key `width`, found `155px`]: for command `assert-css: ("#crate-search", {"width": "159px"})`
/checkout/tests/rustdoc-gui/toggle-docs.goml toggle-docs... FAILED
/checkout/tests/rustdoc-gui/toggle-docs.goml toggle-docs... FAILED
[ERROR] `tests/rustdoc-gui/toggle-docs.goml` line 11: The following errors happened (for selector `#main-content > details.top-doc > summary`): [expected `"Fira Sans", Arial, NanumBarunGothic, sans-serif` for key `font-family`, found `"Source Serif 4", NanumBarunGothic, serif`]: for command `assert-css: (
    "#main-content > details.top-doc > summary",
    {"font-family": '"Fira Sans", Arial, NanumBarunGothic, sans-serif'},

Error: ()
Build completed unsuccessfully in 0:05:18
  local time: Sun Jan 19 11:54:01 UTC 2025
```

</details>

GuillaumeGomez committed:
Add SemiBold for SourceSerif4

GuillaumeGomez commented:
CI fixed, let's go. :)

@bors r=rustdoc

bors commented:
:pushpin: Commit 2511faf61d82dd32fa40f1724e6ee74d51700b54 has been approved by `rustdoc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=rustdoc 2511faf61d82dd32fa40f1724e6ee74d51700b54 -->
<!-- homu: {"type":"Approved","sha":"2511faf61d82dd32fa40f1724e6ee74d51700b54","approver":"rustdoc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
